diverse commits

peter1000 opened the pull request.

MichaelHatherly commented on src/Render/Render.jl:
```diff
@@ -7,4 +7,33 @@ Render
+            c in replace_chars ? write(io, "_") : write(io, lowercase(string(c)))
+        end
+        return takebuf_string(io)
+    end
```
Since this method might get called quite often for the same node, it would be good to cache the generated ``:outname`` before returning it so it only has to be done once.

MichaelHatherly commented on src/Render/Render.jl:
```diff
@@ -7,4 +7,33 @@ Render
+end
+
+function checkchildren(n::Node)
+    isempty(n.children) && throw(EmptyRenderError(string("'$(typeof(n))': '",
```
``typeof(n)`` could be avoided by adding a typevar, ``checkchildren{T}(n::Node{T})``, to the method's signature. Then using ``T`` instead.

peter1000 replied:
thanks for the tip - I really appreciate such.

MichaelHatherly commented on src/Render/Render.jl:
```diff
@@ -7,4 +7,33 @@ Render
+function checkchildren(n::Node)
+    isempty(n.children) && throw(EmptyRenderError(string("'$(typeof(n))': '",
+                                haskey(n.data, :title) ?  n.data[:title] : n.data[:outname],
+                                "' -> has no content")))
```
Do we need `->` here. It always makes me immediately think of anonymous functions, which clearly isn't what the error message is about. Just

```
"' has no content."))
```

would be fine.

peter1000 commented:
* Empty node: do you think empty nodes should be allowed - to me it seems it does not make much sense? Just asking as in Render I wanted to check that - but maybe this could be done already in the Elements part?

* You mentioned about caching the `:outname`. In Render I wanted to require either an `:outname` or a `:title` where a outname is generated from.

I was thinking if also such should not already be done in Elements?

`:outname` I thought to have to allow for something like:

page(
      mods.....,
      title = "Longe unicode title - 疯狂寿命长标题",
      outname = "favorites"
)

I guess in many cases the `title` will be enough but for certain cases an manual entered outname might be appropriated too.

My thought was to check in Elements for childless nodes and if no `:outname` key is in the Dict - generate one from `title` and add the key. else throw an error?

peter1000 commented:
For empty node check add a small line to `build` something like: 

```julia
function build!{T}(node::Node{T}, args, kwargs)
    isempty(args) && throw(ArgumentError("Empty '$(T)'."))
    for arg in args update!(node, arg) end
    update!(node, kwargs)
    node
end
```

A number of tests I would need to update.

peter1000 commented:
I change the PR to your suggestion but added it to Elements - seems to me it looks much cleaner. I will update the tests to reflect the changes.

If you do not like the empty node check - well than it was an exercise for me.

peter1000 commented:
This might fail cause Fact pacht migth be missing - any way just for a review if that would be not the better way than my first try. I can Comment out the couple of failing test till Fact is patched?

peter1000 commented:
don't understand why winows failed?

peter1000 commented:
today I plan to push here a few commits more

peter1000 commented:
[coverage line ](https://coveralls.io/builds/2688241/source?filename=src%2FElements%2FElements.jl#L45)  when is that called?

Do you have a test for it?

MichaelHatherly commented:
> coverage line when is that called?

When two ``Node`` objects have different parameters. There's no test for that just yet.

> Any idea what's wrong with Windows failed test?

Not sure about that one. It looks like it might be to do with line endings in windows, ``\r\n`` instead of ``\n`` or something like that.

I'm not a fan of throwing an error on empty nodes really. Its going to get annoying when someone is starting to write their documentation an has to put in placeholder text if they haven't yet got content for a specific ``page``.

The other error type names are exposing internals to the user, they don't really need to know anything about children or parent nodes.

peter1000 commented:
I gess the error: `ERROR: LoadError: LoadError: LoadError: LoadError: UndefVarError: MarkdownFormatter not defined` is because the lates Docile changes are not tagged ?

peter1000 commented:
Hi,

I uploaded a first test docile documentation: still a lot missing like proper signatures only use the name for now 

[first new lexicon docs](http://peter1000.github.io/new_lexicon_example_docs1/)

The used [build.jl](https://github.com/peter1000/Lexicon.jl/blob/diverse_commits/examples/docile_documentation/build.jl) writes also the `mkdocs.yml`   I hope this is more like what you had in plan.

- check out the permalinks ect..

peter1000 commented:
Next plan: 
- working on the signatures

- sorter for `docs`  **maybe you could give me some examples** what you had in mind like you did with the filters?

MichaelHatherly commented on src/Elements/utilities.jl:
```diff
@@ -0,0 +1,35 @@
+["Helper functions."]
+
+"""
+Find the final configuration fo node `n`.
```
For code syntax in markdown can we consistently use double backticks. It's much more noticeable when reading and I've been trying to change them in the codebases when updating surrounding code.

peter1000 replied:
no problem - I will skip through the code and change them. Done

MichaelHatherly commented on src/Elements/utilities.jl:
```diff
@@ -0,0 +1,35 @@
+Search begins with the node `n` itself followed by any parent `Node` configuration setting.
+"""
+function findconfig(n::Node)
+    local config
```
Not needed, since it's immediately assigned to.

peter1000 replied:
Thanks - left over as I tried first something else.

MichaelHatherly commented on src/Elements/utilities.jl:
```diff
@@ -0,0 +1,35 @@
+        update!(config, n.data)
+    end
+    config
+end
```
``findconfig`` shouldn't be mutating any of the nodes' ``config``s. Just search for the closest config and return that. Or even better, require a ``Symbol`` arg to ``findconfig(n::Node, field::Symbol)`` and just return the field's value.

Basically, once created the ``document(...)`` object shouldn't be changed, just used for lookup otherwise it's going to get difficult to debug if it's getting changed internally.

peter1000 replied:
I thought about that: but my conclusion was that if one looks for a couple of symbols it has always to rerun all.

* I had a version with the symbol lookup.

  `getconfig(n::Node, k:Symbol)` but was unsure what to do if the key is not found. If we throw an error we always have to account for that in the code?

 I thought maybe return `nothing` and in the code? Or better use `try` ?

  ```
  f = getconfig(docs, :filter)
  f != nothing && ....
  ```

* > Basically, once created the document(...) object shouldn't be changed

  Do avoid re-run should we catch the final config when created the document(...) ?

MichaelHatherly commented on src/Elements/utilities.jl:
```diff
@@ -0,0 +1,35 @@
+
+"""
+Adds a new key ``k`` with value ``v`` to the ``node`` data dictionary .
+Throws an ArgumentError if the key exists.
```
"ArgumentError" -> "\`\`ArgumentError\`\`".

peter1000 replied:
Done

MichaelHatherly commented on src/Render/Render.jl:
```diff
@@ -5,6 +5,19 @@ module Render
 using Compat
 
+include("utilities.jl")
+include("md.jl")            # Markdown rendering.
```
"md" -> "markdown"

peter1000 replied:
Done

MichaelHatherly commented on src/Render/md.jl:
```diff
@@ -0,0 +1,120 @@
+type RenderedMarkdown
+    outdir      :: UTF8String
+    outpages    :: Vector{Tuple{AbstractString, AbstractString}}
+    layout      :: Vector{Tuple{UTF8String, Vector}}
```
Would these two ``Vector{Tuple{...}}`` be better as ``Dict{String, String}``? Do they need to be ordered?

peter1000 replied:
`outpages` I think could be a dict: but must implement it later and see if it actually works the same.

The other one has to stay ordered.  Anyway I mentioned for you [here](https://github.com/MichaelHatherly/Lexicon.jl/issues/112#issuecomment-108263471) that they plan to implment a OrderedDict / or replace the existing one in 0.4 release.

MichaelHatherly commented on src/Render/md.jl:
```diff
@@ -0,0 +1,120 @@
+        objname = nameof(m, obj)    # Replace that later
+        println(io, "---\n")
+        #println(io, string(config.style_obj, " ",  objname))
+        println(io, "#### $(objname)")
```
This should be configurable?

peter1000 replied:
yes just a dummy for now: till the signature output is working nicely.

MichaelHatherly commented on src/Render/md.jl:
```diff
@@ -0,0 +1,120 @@
+        else
+            throw(ArgumentError(string("\n'save(): outdir' exists. `remove_destination=true` ",
+                                       "is required to remove it before saving.\n",
+                                       "outdir: $(rmd.outdir)\n\n")))
```
Triple quoted string.

peter1000 replied:
In Base they seem to use it mostly like [this](https://github.com/JuliaLang/julia/blob/master/base/file.jl#L83-92)

Anyway changed it. - not sure if you meant to move all to the line front?

```julia
else
    throw(ArgumentError(string("""
       ``outdir`` exists. 
       ``remove_destination=true`` is required to remove it before saving.
       outdir: $(rmd.outdir)
       """)))
end
```

MichaelHatherly commented on src/Utilities.jl:
```diff
@@ -5,6 +5,66 @@ module Utilities
+nameof(s::Symbol) = s
+nameof(m::Module, obj::Symbol)            = obj
+nameof(m::Module, obj::Module)            = module_name(obj)
+nameof(m::Module, obj::Method)            = nameof(Cache.getmeta(m, obj)[:code])
```
``nameof(m::Module, obj::Method) = obj.func.code.name``?

peter1000 replied:
seems that method is supposed to do what I indented. [name and module reflection](https://github.com/JuliaLang/julia/blob/master/base/reflection.jl#L4)

Leave it for now.

MichaelHatherly replied:
> seems that method is supposed to do what I indented.

Which method are you referring to? I'm not talking about ``nameof(m::Module, obj::Module)`` but ``nameof(m::Module, obj::Method)`` rather.

peter1000 replied:
sorry I misread that - will change it.

MichaelHatherly commented on src/Utilities.jl:
```diff
@@ -5,6 +5,66 @@ module Utilities
+nameof(m::Module, obj::Module)            = module_name(obj)
+nameof(m::Module, obj::Method)            = nameof(Cache.getmeta(m, obj)[:code])
+nameof(m::Module, obj::DataType)          = obj.name.name
+nameof(m::Module, obj::QualifiedSymbol)   = symbol(last(split(string(obj), '.')))
```
``nameof(m::Module, obj::QualifiedSymbol) = obj.sym``?

peter1000 replied:
Changed - Thanks

MichaelHatherly commented on src/Utilities.jl:
```diff
@@ -5,6 +5,66 @@ module Utilities
+        nameof(x.args[2]) :
+        isexpr(x.args[1], :(.)) ?
+                      x.args[1] :
+                      nameof(x.args[1])
```
```jl
function nameof(x::Expr)
    isa(x.args[1], Bool)    && return nameof(x.args[2])
    isexpr(x.args[1], :(.)) && return x.args[1]
    nameof(x.args[1])
end
```

seems clearer perhaps? If I'm reading that chain correctly...

peter1000 replied:
Yes looks nicer ;) changed

I took that from [your other code](https://github.com/MichaelHatherly/Docile.jl/blob/master/src/Legacy/atdoc.jl#L40-46)

Maybe this should also be changed or just leave it as it is the legacy code and might be removed in the future?

MichaelHatherly commented on src/Utilities.jl:
```diff
@@ -5,6 +5,66 @@ module Utilities
+    elseif meta[:category] == :macro
+        symbol(string("@", nameof(meta[:code])))
+    else
+        throw(ArgumentError("'$obj' wrong category: $(meta[:category])"))
```
So long as Docile's doing the right thing and nothing is being mutated then this shouldn't be needed? Have you found examples to the contrary?

peter1000 replied:
I thought so, but was unsure and just added it to find any exceptions.

Will remove it.

MichaelHatherly commented on src/Utilities.jl:
```diff
@@ -5,6 +5,66 @@ module Utilities
+end
+
+"""
+Is the documented object ``obj`` been exported from the given module ``m``?
```
Either change "Is" to "Has" or remove "been". I'd go with present tense for docs generally and remove "been".

peter1000 replied:
Done

MichaelHatherly commented on src/Utilities.jl:
```diff
@@ -5,6 +5,66 @@ module Utilities
+"""
+Is the documented object ``obj`` been exported from the given module ``m``?
+"""
+isexported(m::Module, obj) = nameof(m, obj) in names(m)
```
Each ``ModuleData`` object stores the exports for the module in ``metadata[:exports]``. Although longer, ``Cache.getmeta(Cache.getmodule(m))[:exports]``, doesn't need to recompute ``names`` each time.

Since it's a ``Set`` rather than ``Vector`` ``in`` *should* be faster for larger numbers of exports. Less allocations as well I'd think.

MichaelHatherly commented on src/Utilities.jl:
```diff
@@ -5,6 +5,66 @@ module Utilities
+
+"""
+Is the object ``obj`` from module ``m`` a Docile category ``cat`` or one of the categories ``cats``.
+"""
```
Since you're describing both methods here it should document the ``Function`` rather.

peter1000 replied:
Change to:

```julia
"""
Is the object ``obj`` from module ``m`` a Docile category ``cat`` or one of the categories ``cats``.
"""
:iscategory
``` 

hoe this is correct

MichaelHatherly commented:
Regarding the "examples" directory:

I'd rather we just provide a link in the Lexicon docs to known ``build.jl`` scripts once they're written, such a Lexicon's and Docile's own ones. Examples tend to end up out of date sooner or later and it's less to have to maintain.

peter1000 commented:
> Regarding the "examples" directory: Examples tend to end up out of date sooner or later and it's less to have to maintain.

you are right. could we just keep it for now - so we can more easily refer to the same thing?

peter1000 commented on examples/docile_documentation/build.jl:
```diff
@@ -0,0 +1,82 @@
+                "# $m",
+                "## External",
+                "### Modules, Functions, Methods & Macros",
+                docs(m, filter = obj -> iscategory(m, obj, [:module, :function, :method, :macro]) && isexported(m, obj)),
```
Maybe you have an idea how to do this.

I would like to only write:

`iscategory(obj, [:module, :function, :method, :macro]), isexported(obj)),` withouth the additional module reference.

The code it calls is [here](https://github.com/MichaelHatherly/Lexicon.jl/pull/116/files#diff-01285fe4a25bbc4da684aa18148e934eR72)

```julia
function innerpage!(io::IO, n::Node, m::Module, emptyline::Bool)
    warn("!! TODO: `Page Docs Module` not finished yet!!")
    objects = Cache.objects(m)
    # get any filter config for this node
    conf = findconfig(n)
    objs = haskey(conf, :filter) ? filter(conf[:filter], objects) : objects

    for obj in objs
    .....
```

You also had in mind to change `conf = findconfig(n)` to return the value for `conf = findconfig(n, :filter)`

peter1000 commented:
**Next plan:**

``outpages`` try a Dict instead of Vector{Tuple} [comment](https://github.com/MichaelHatherly/Lexicon.jl/pull/116#discussion_r31617664)

continue working on the signatures

Question:  `sorter` for docs maybe you could give me some examples what you had in mind like you did with the filters?

peter1000 commented:
Maybe you have an idea how to do this.

I would like to only write:

`iscategory(obj, [:module, :function, :method, :macro]), isexported(obj)),` withouth the additional module reference.

```julia
pages  = [page(
            "# $m",
            "## External",
            "### Modules, Functions, Methods & Macros",
            docs(m, filter = obj -> iscategory(m, obj, [:module, :function, :method, :macro]) && isexported(m, obj)),
            title = "$m")

            for m in [
                Docile,
                Cache,
                ....
                ]
        ]
````
The code it calls is [here](https://github.com/MichaelHatherly/Lexicon.jl/pull/116/files#diff-01285fe4a25bbc4da684aa18148e934eR72)

```julia
function innerpage!(io::IO, n::Node, m::Module, emptyline::Bool)
    warn("!! TODO: `Page Docs Module` not finished yet!!")
    objects = Cache.objects(m)
    # get any filter config for this node
    conf = findconfig(n)
    objs = haskey(conf, :filter) ? filter(conf[:filter], objects) : objects

    for obj in objs
    .....
```

You also had in mind to change `conf = findconfig(n)` to return the value for `conf = findconfig(n, :filter)`

MichaelHatherly commented:
> Maybe you have an idea how to do this.

Not sure what you'd like to be doing here. I believe the module argument is necessary unless I'm mistaken.

peter1000 commented:
> I believe the module argument is necessary unless I'm mistaken.

Yes at the moment. Could we add a reerence to the module to all obj.  little pit like the Elements parents reference?

MichaelHatherly commented on src/Elements/utilities.jl:
```diff
@@ -0,0 +1,24 @@
+Return the value stored for the given key, or the given default value if no mapping for the key is present.
+Search begins with the node ``n`` itself followed by any parent node configuration setting.
+"""
+function getconfig(n::Node, key::Symbol, default::Any=:notfound)
```
``::Any`` isn't needed in the signature. Spaces between ``default`` and ``:notfound`` to follow the rest of the codebase.

peter1000 replied:
done

MichaelHatherly commented on src/Elements/utilities.jl:
```diff
@@ -0,0 +1,24 @@
+"""
+function getconfig(n::Node, key::Symbol, default::Any=:notfound)
+    haskey(n.data, key) && return n.data[key]
+    # Stage 2: Parent's config.
```
Unnecessary comment perhaps.

peter1000 replied:
removed

MichaelHatherly commented on src/Externals/mkdocs.jl:
```diff
@@ -1 +1,72 @@
+    [haskey(conf, k) ||
+        throw(ArgumentError("Document '$(conf[:title])' has no config key ':$k'.")) for k in keys]
+
+    [haskey(conf, k) && push!(keys, k) for k in optionalkeys]
```
These unassigned comprehensions confused me for a moment. Neat though! I'm slightly apprehensive about them though, could be clearer with plain old ``for``-loops?

The comprehensions are allocating arrays, though not a high priority for this bit of code I'd think. Something to keep in mind.

peter1000 replied:
changed

MichaelHatherly commented on src/Render/markdown.jl:
```diff
@@ -0,0 +1,122 @@
+    outdir      :: UTF8String
+    outpages    :: Dict{AbstractString, AbstractString}
+    layout      :: Vector{Tuple{UTF8String, Vector}}
+    document    :: Node{Document}
```
Too many spaces.

peter1000 replied:
adjusted

MichaelHatherly commented on src/Render/markdown.jl:
```diff
@@ -0,0 +1,122 @@
+    return RenderedMarkdown(abspath(outdir), outpages, layout, document)
+end
+
+function save(rmd::RenderedMarkdown; remove_destination=false)
```
Single spacing around the ``=``.

peter1000 replied:
done

MichaelHatherly commented on src/Render/markdown.jl:
```diff
@@ -0,0 +1,122 @@
+    # mkdocs yaml
+    haskey(rmd.document.data, :mkdocsyaml) && rmd.document.data[:mkdocsyaml] && writemkdocs(rmd)
+    return rmd
+end
```
I'm sure some of this file's contents would be applicable to other formats, yes? Don't have to worry about that just yet, but worth keeping in mind.

peter1000 replied:
ok

MichaelHatherly commented:
> Could we add a reference to the module to all obj.

You can get the module reference for any of the objects already from the objects themselves. The only ones that were troublesome were ``:global`` and ``:typealias``, which are fixed with ``QualifiedSymbol`` now. Which ones are giving trouble?

MichaelHatherly commented:
Docile ``0.5.4`` will have ``MarkdownFormatter``, busy tagging the version now.

peter1000 commented:
> You can get the module reference for any of the objects already from the objects themselves.  Which ones are giving trouble?


I got this

```julia
getmod(obj::Module)          = obj
getmod(obj::Method)          = obj.func.code.module
getmod(obj::DataType)        = obj.name.module
getmod(obj::QualifiedSymbol) = obj.mod

```

I got a bit stuck here:

- Aside
- function

MichaelHatherly commented:
For ``Function`` just get the module of the first ``Method`` in it's method table.

peter1000 commented:
> For Function just get the module of the first Method in it's method table.

sorry I'm not sure how you want to do that if there is no method specified?

if I have

```
"f_1"
function f_1 end
```

peter1000 commented:
That is what I have till now:

```julia

getmod(obj::Module)          = obj
getmod(obj::Method)          = obj.func.code.module
getmod(obj::DataType)        = obj.name.module
getmod(obj::QualifiedSymbol) = obj.mod
getmod(obj::Aside)           = obj.mod
function getmod(obj::Function)
    if isgeneric(obj)
        isa(obj.env.defs, Void)                ? 
            error("TODO: getmod for function") : 
            obj.env.defs.func.code.module
    else # Macro
        obj.code.module
    end
end
```

QUESTION: is that check for the macro ok?

Of course it is not usable because it breaks by a `only function` like I mentioned above

peter1000 commented:
Found some more problems with getting the first method.

I defined a function in Docile.Cache

```julia
"do_something"
function do_something end
```

and in Docile.Extension

```julia
do_something() = return ""
```

function module: Docile.Extension

peter1000 commented:
Here is my thought: would it not be easier to have an additional ObjectIdDict for the mod.

**Advantage** I can think of: 

- can get rid of the `Qualified Symbols` all togehter, can't we?
- can get rid of the `Aside mod` field I planned to merge
- can get rid of all the ``getmod`` functions I wanted to add and have a single:

    getmod(obj)

- filters do not need to add the mod
    
  `docs(Docile.Cache, filter = obj -> iscategory(Docile.Cache, obj, [:macro]) && isexported(Docile.Cache, obj)),`

  can be written like
    
  `docs(Docile.Cache, filter = obj -> iscategory(obj, [:macro]) && isexported(obj)),`


**Disadvantage**

- not sure maybe it's a bit slower to look up than `obj.func.code.module`  ?
- one more ObjectIdDict but no Oulified Symbols etc..


[DocsCache](https://github.com/MichaelHatherly/Docile.jl/blob/master/src/Cache/types.jl#L6)

"""
For a single module store raw docstrings, parsed docs, and metadata.
"""
type DocsCache
    raw    :: ObjectIdDict
    parsed :: ObjectIdDict
    meta   :: ObjectIdDict
    mod    :: ObjectIdDict   # ADD THIS ??

    DocsCache(raw, meta) = new(raw, ObjectIdDict(), meta, mod)
end

MichaelHatherly commented:
``function f end`` isn't too much of a priority at the moment since it's extremely new, also in real code you'd have defined at least one method on the function somewhere.

> Found some more problems with getting the first method.

Those are two distinct ``Function`` objects, not the same ones.

> Here is my thought: would it not be easier to have an additional ObjectIdDict for the mod.

A ``DocsCache`` is generated per-``Module`` so a ``mod`` field would just contain references to the same module for each object; not an improvement really.

I'm not sure how you're ``filter -> iscategory(obj, [:macro]) && isexported(obj)`` could work? What if ``obj`` is a ``Symbol``?

peter1000 commented:
> Those are two distinct Function objects, not the same ones.

Not sure but I documented only the one and  `Cache.objects(Cache)`  returned "Docile.Extension"?

> I'm not sure how you're filter -> iscategory(obj, [:macro]) && isexported(obj) could work? What if obj is a Symbol?

I would have to try that - if you say so probably I'm wrong ;)

MichaelHatherly commented:
> Not sure but I documented only the one and ``Cache.objects(Cache)`` returned "Docile.Extension" as getmod for the obj

Then I assume you imported the method into ``Docile.Extensions`` to extend it rather? Difficult to tell without a full example though.

peter1000 commented:
> I'm not sure how you're filter -> iscategory(obj, [:macro]) && isexported(obj) could work?

Seems to work for all the Docile Modules

```julia

isexported(obj) = nameof(getmod(obj), obj) in Cache.getmeta(Cache.getmodule(getmod(obj)))[:exports]

iscategory(obj, cat::Symbol)          = Cache.getmeta(getmod(obj), obj)[:category] == cat
iscategory(obj, cats::Vector{Symbol}) = Cache.getmeta(getmod(obj), obj)[:category] in cats

```

Build code

```julia

pages  = [page(
            "# $m",
            "## External",
            "### Modules, Functions, Methods & Macros",
            docs(m, filter = obj -> iscategory(obj, [:macro]) && isexported(obj)),
            "### Globals, Types, Asides",
            docs(m, filter = obj -> iscategory(obj, [:aside]) && isexported(obj)),
            "## Internal",
            "### Modules, Functions, Methods & Macros",
            docs(m, filter = obj -> iscategory(obj, [:macro]) && !isexported(obj)),
            "### Globals, Types, Asides",
            docs(m, filter = obj -> iscategory(obj, [:aside]) && !isexported(obj)),
            title = "$m")

            for m in [
                Docile,
                Cache,
                Collector,
                Extensions,
                Formats,
                Interface,
                Legacy,
                Runner,
                Utilities,
                ]
        ]
```

peter1000 commented:
> function f end isn't too much of a priority

Just to be consistent maybe we could disable documenting a single function. and just silently drop it here?

Something like

```julia
function getmod(obj::Function)
    if isgeneric(obj)
        isa(obj.env.defs, Void) || obj.env.defs.func.code.module
    else # Macro
        obj.code.module
    end
end
```

BTW: is this correct for macro check? `!isgeneric`

MichaelHatherly commented:
> Seems to work for all the Docile Modules

The thing is that you can actually document the same object in different modules so the module argument is needed really. I know it's very much a corner case, but I'd prefer not to paint ourselves into a corner by removing the argument. See the following example:

```jl
module Outer

module Inner

"inner"
Inner

end

"outer"
Inner

end
```

> Just to be consistent maybe we could disable documenting a single function.

Do you mean a ``Function`` with 0 methods?

> BTW: is this correct for macro check? ``!isgeneric``

I think so. There's some code in Docile that does a similar thing, in the ``includefiles`` function.

> Do you think we should put this ``getmod()`` functions better in Docile

In Docile rather, if we actually need this.

> and rename the nameof to name and put it also into Docile - you had a couple of them there in the legacy see or move all to Lexicon ??

They'd all need to deprecated until Docile 0.6, so perhaps more hassle than it's worth moving them over.

peter1000 commented:
> Do you mean a Function with 0 methods?

yes I mean that.

> The thing is that you can actually document the same object in different modules so the module argument is needed really.

I don't understand this - sorry. If one get's the module internally from a obj (which seems to work except for a  Function with 0 methods (which we might sidallow to be documented) than it seems I do not need the mod argument. But with multiple dispatch we can easily provide both options can't we?

peter1000 commented:
>>    Do you think we should put this getmod() functions better in Docile

>> In Docile rather, if we actually need this.

So can I move thsi to Docile.Utilities / or Interface

```julia
getmod(obj::Module)          = obj
getmod(obj::Method)          = obj.func.code.module
getmod(obj::DataType)        = obj.name.module
getmod(obj::QualifiedSymbol) = obj.mod
getmod(obj::Aside)           = obj.mod

function getmod(obj::Function)
    if isgeneric(obj)
        isa(obj.env.defs, Void)                                                                   ?
            throw(ArgumentError("At least one method must be defined for a documented function")) :
            obj.env.defs.func.code.module
    else # Macro
        obj.code.module
    end
end

```

`iscategory`, `isexported` we leave in Lexicon I suppose?

peter1000 commented:
I will add them better to: `docile.Cache.interface`

peter1000 commented:
> Do you mean a Function with 0 methods?

yes I meant that. But if we require the `mod` it doesn't matter and probably should stay as it is until we run into troubles.

> The thing is that you can actually document the same object in different modules so the module argument is needed really.

Ok. So we leave it as it is. I will revert all back.

peter1000 commented:
Reverted PR back and go from here.

Question: if we require anyway everwhere the module: are the Qualified Symbols really needed?

MichaelHatherly commented:
Have I missed something in this issue? Seems some messages have disappeared...

peter1000 commented:
> Seems some messages have disappeared...

I deleted them after I reverted to the previous commit. I had posted a couple of messages regarding `getmod` which we decided to remove and require always that the module is also passed. 

I still would be interested to know what you had in mind about: `sorter` - just something similar to `filter` ? any examples?

MichaelHatherly commented:
> I deleted them after I reverted to the previous commit. I had posted a couple of messages regarding getmod which we decided to remove and require always that the module is also passed.

Ok.

> I still would be interested to know what you had in mind about: sorter - just something similar to filter ? any examples?

Perhaps a similar interface to what ``Base.sort`` does with the ``lt = <comparision>`` keyword argument.

peter1000 commented:
#### UPDATED EXAMPLE DOCS

Note: with the new mkdocs also the copyright footer works for the readthedoc theme

Note: tried default style based on [python with](https://docs.python.org/3/library/functions.html#globals)

- bold name
- italic arguments

  ![screenshot from 2015-06-09 10-06-17](https://cloud.githubusercontent.com/assets/6168860/8058381/693a1ef0-0e8f-11e5-8fb0-5a0d4c14538b.png)

Have a look [example Docs](http://peter1000.github.io/new_lexicon_example_docs1/api/docile.collector/#modules-functions-methods-macros_1)

about the [txt][src] would probably nicer if they are also only show if one hoovers over it like the permalink ?

peter1000 commented:
BTW: I found a small annoying thing because we require the module to get the meta.

I added an `obj_style` function: I first wanted to add only one to a section but because of the module requirement one has to add it to each `docs()`

MichaelHatherly commented:
Would it be possible to squash some of these commits? Also might be best to submit some smaller PRs for independant pieces of this PR separately first, such as the mkdocs ``Externals``. It'll make it much easier for me to review things.

peter1000 commented:
> Would it be possible to squash some of these commits? Also might be best to submit some smaller PRs for independant pieces

I thought so - I think I just want to do some more work on the signature and methods definitions and than split it piece for piece into smaller PR we can actually merge. Because of some personal changes I do not have anymore time in the evening/night so it might take me a couple of days.

MichaelHatherly commented:
Sure, no hurry.

MichaelHatherly commented:
I've pushed some REPL rendering code that gets us some basic help in the terminal. There's couple of pieces I needed from this PR that are now on ``lexicon-next`` branch for you to use in this PR once you rebase.

MichaelHatherly commented:
I want to have most of the sorting/filtering/parsing done only once rather than have it as part of the final rendering stage (to markdown, LaTeX, etc). I've just push (ed9bd2ea1cc45a21d4087e9fb98c2daed246e473) code for "stage 2" of the rendering, ie. compiling and caching some of the work needed that doesn't involve actually outputting any documentation.

The three stages should then be:

**Stage 1:**

```jl
out = document(
    section(...),
    section(...),
)
```

Capture the structure and configuration for the documentation tree in ``out``. Configuration unrelated to the output formats should be specified here.

**Stage 2:**

```julia
compile!(out)
```

Group, sort, and filter the docstrings. Read all file paths and parse strings into markdown (or other) AST. Generate IDs for each node. Anything else you think should happen at this point?

**Stage 3:**

```julia
save(out, Markdown(<<config>>))
save(out, LaTeX(<<config>>))
```

Take the documentation tree ``out`` and render it using the provided format, ``Markdown``, ``LaTeX``, etc that provides any format-specific configuration needed. ``save`` will call ``compile!`` internally so the user doesn't have to worried about that step.

I don't think stage 3 should be mutating anything in the ``.cache`` fields since that might result in the order of calls to ``save`` mattering. State for stage 3 should be stored in the ``Markdown`` object passed as the second object to ``save``.

peter1000 commented:
I think that general the 3 stages are a good idea. I'm a bit unsure about the fromat specific configuration.

If all config at the `out level 1 stage` must be of general nature some of the power of the configuration system might be gone but it surely makes it easier to output the same `document` to different formats.

Just an example from the top of my head:

---

html for instance supports colors - which plain markdown does not: so maybe someone wants a header formated with a `color = blue` config in the `out 1 level stage`

I'm not sure passing this format specific only to stage 2 will be as flexible and easy to handle. I thought that was one reason you choose the `config` to be a Dictionary.

MichaelHatherly commented:
> html for instance supports colors - which plain markdown does not: so maybe someone wants a header formated with a color = blue config in the out 1 level stage

Seems like something to be done with css rather, ie. styling ``h1`` instead of hard-coding a color? And markdown does allow for arbitrary raw html to be used inline as well. Perhaps we could accept like Hiccup nodes in place of plain strings.

peter1000 commented:
> Seems like something to be done with css rather, ie. styling h1 instead of hard-coding a color?
yes just as an example: in html one might want to set a class for styling `class="special"`

But other formats might require other options.

Anyway: both variations have advantages and if one wants to output into multiple formats probably a more generic approach is best.

> Hiccup nodes

never heard of that sorry

peter1000 commented:
A other thought: If one wants the config to be useful across all formats: I think one has to predefine supported config keys.

MichaelHatherly commented:
> never heard of that sorry

https://github.com/one-more-minute/Hiccup.jl

> I think one has to predefine supported config keys.

Yes, we can certainly look into doing that.

peter1000 commented:
> Hiccup.jl

Maybe it's better to wait until some general things are working - than one might see easier if it is needed.

peter1000 commented:
hi,

* just saw this and thought I put it here as info.
  [This breaks some packages' uses of FactCheck](https://github.com/JuliaLang/julia/pull/12285#issuecomment-124782952)

* What do you think about the status of the doc issue in julia.
  Do you think it is fine to proceed here where we left of about a month ago (how time flies) or better to still wait?

MichaelHatherly commented:
> This breaks some packages' uses of FactCheck

From the test runs this morning it appears to not affect either Docile or Lexicon, I think.

> Do you think it is fine to proceed here where we left of about a month ago (how time flies) or better to still wait?

Up to you really. I've got minimal free time for anything other than quick maintenance fixes at the moment, but will happily review anything you'd like to work on. I'm sure there will still be room for Docile/Lexicon in the ecosystem after 0.4 is released, but probably a much reduced number of potential users.

peter1000 commented:
Thanks for the feedback.

I'm actually not uptodate with the doc changes within juia - based on your comment I think I will wait till 0.4 is out.

I close this in the mean time as that PR is any completely out of sync
